File: pubnub-gson/pubnub-gson-api/src/main/java/com/pubnub/api/java/models/consumer/datasync/relationship/PNDataSyncRelationship.java
package com.pubnub.api.java.models.consumer.datasync.relationship;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import lombok.experimental.Accessors;
import org.jetbrains.annotations.Nullable;

import java.util.Map;

/**
 * DataSync Relationship resource returned by {@code pubnub.dataSync().getRelationship()} /
 * {@code createRelationship()}.
 *
 * <p>A Relationship links two entities (entity A and entity B) under a relationship class. It is the general
 * counterpart to Membership, which is a fixed Channel↔User specialization.
 */
@Getter
@Setter
@NoArgsConstructor
@Accessors(chain = true)
@ToString
public class PNDataSyncRelationship {
    /**
     * Relationship identifier.
     */
    private String id;

    /**
     * Identifier of entity A this relationship links.
     */
    private String entityAId;

    /**
     * Identifier of entity B this relationship links.
     */
    private String entityBId;

    /**
     * Relationship class identifier.
     */
    private String className;

    /**
     * Version of the relationship class.
     */
    private int classVersion;

    /**
     * Date and time the relationship was created.
     */
    private String createdAt;

    /**
     * Date and time the relationship was last updated.
     */
    private String updatedAt;

    /**
     * The relationship's content fingerprint used in conditional requests.
     */
    private String eTag;

    /**
     * Relationship status.
     */
    @Nullable
    private String status;

    /**
     * Date and time when the relationship expires (will be deleted automatically). Always present:
     * server-computed on every response, never client-settable.
     */
    private String expiresAt;

    /**
     * Arbitrary user-defined JSON object.
     */
    @Nullable
    private Map<String, Object> payload;
}

File: pubnub-kotlin/pubnub-kotlin-api/src/commonMain/kotlin/com/pubnub/api/models/consumer/datasync/relationship/PNDataSyncRelationship.kt
package com.pubnub.api.models.consumer.datasync.relationship

import com.pubnub.api.utils.SerializedName

/**
 * DataSync Relationship resource returned by `pubnub.dataSync.getRelationship`, `createRelationship` and
 * referenced by remove operations. A Relationship is a typed link between two entities (side A and side B) under
 * an arbitrary relationship class, exposed under the generic two-end names [entityAId] / [entityBId].
 *
 * @property id Relationship identifier.
 * @property entityAId Identifier of entity A this relationship links.
 * @property entityBId Identifier of entity B this relationship links.
 * @property className Relationship class identifier.
 * @property classVersion Version of the relationship class.
 * @property createdAt Date and time the relationship was created.
 * @property updatedAt Date and time the relationship was last updated.
 * @property eTag The relationship's content fingerprint used in conditional requests.
 * @property expiresAt Date and time when the relationship expires (will be deleted automatically). Always
 * present: server-computed on every response, never client-settable.
 * @property status Relationship status.
 * @property payload Arbitrary user-defined JSON object.
 */
data class PNDataSyncRelationship(
    val id: String,
    val entityAId: String,
    val entityBId: String,
    @field:SerializedName("relationshipClass") val className: String,
    @field:SerializedName("relationshipClassVersion") val classVersion: Int,
    val createdAt: String,
    val updatedAt: String,
    val eTag: String,
    val expiresAt: String,
    val status: String? = null,
    val payload: Map<String, Any?>? = null,
)
